### PR TITLE
Support multi-lines matching for function

### PR DESCRIPTION
**Pull request by littleGnAl**
Introduce `LanguageSyntaxMatcher.findFunctionNameFromBlock` to allow matching the function name from the function block, so that you can fine-grained control what the function name you want from the code block.

### CODE REVIEW

File: iris_doc/api_tagger.py [hunk 1]
```diff
@@ -254,13 +271,17 @@ def _findClassScopeEndIndex(self, lines: List[str], scopeStartIndex: int) -> int
...
+        while (i < len(lines)):
             line = lines[i].strip()
-            if self.__syntaxMatcher.matchClassScopeEnd(line) or self.__syntaxMatcher.matchFunctioinScopeEnd(line):
+            if self.__syntaxMatcher.matchClassScopeEnd(line) or \
+                    (has_match_function_start and self.__syntaxMatcher.matchFunctioinScopeEnd(line)):
                 scopeStack.pop()
+                has_match_function_start = False
             elif self.__syntaxMatcher.matchFunctioinScopeStart(line):
                 scopeStack.append(i)
+                has_match_function_start = True
```
Comment (maxxfrazer) [inline iris_doc/api_tagger.py]: I just tried the following in my branch, removing the ObjCLineScanner check and it passes the tests in all languages:
```py
        while(i < len(lines)):
            line = lines[i].strip()
            if self.__syntaxMatcher.matchFunctionScopeStart(line) or self.__syntaxMatcher.matchClassScopeStart(line):
                scopeStack.append(i)
            if self.__syntaxMatcher.matchClassScopeEnd(line) or (not self.__syntaxMatcher.matchMemberVariable(line) and self.__syntaxMatcher.matchFunctionScopeEnd(line)):
                scopeStack.pop()
            if len(scopeStack) == 0:
                return i
```

Functions and properties in objc header files end the same way, with a `;`. while classes, enums, prototypes end with `@end`, and do not nest.

File: iris_doc/api_tagger.py [hunk 2]
```diff
@@ -273,7 +294,43 @@ def _findFunctionScopeStartIndex(self, lines: List[str], startIndex: int) -> int
         return self._findClassScopeStartIndex(lines=lines, startIndex=startIndex)
 
     def _findFunctionScopeEndIndex(self, lines: List[str], scopeStartIndex: int) -> int:
-        return self._findClassScopeEndIndex(lines=lines, scopeStartIndex=scopeStartIndex)
+        scopeStack = []
+        line = lines[scopeStartIndex].strip()
+        # For a single line function, return the `scopeStartIndex` directly
+        if self.__syntaxMatcher.matchFunctioinScopeEnd(line) or \
+                self.__syntaxMatcher.matchAbstractFunctioinScopeEnd(line):
+            return scopeStartIndex
```
Comment (maxxfrazer) [inline iris_doc/api_tagger.py:298]: Should this be `and` instead of `or`?

File: iris_doc/api_tagger.py (hunk 1 above)
Comment (littleGnAl) [reply iris_doc/api_tagger.py]: @maxxfrazer Updated.

File: iris_doc/api_tagger.py (hunk 2 above)
Comment (littleGnAl) [reply iris_doc/api_tagger.py:298]: The function scope end and abstract function scope end are not the same in most of languages. I think it should be `or` instead.

Comment (maxxfrazer) [reply iris_doc/api_tagger.py:298]: :+1: I misread the code actually!

**Review by maxxfrazer — APPROVED** [review]
LGTM

**Review by maxxfrazer — APPROVED** [review]

**Review by LichKing-2234 — APPROVED** [review]

**Pull request merged**